Fixes #24336 - upgrade jest and fix errors

johnpmitsch opened the pull request:
This does a few things:
- upgrades jest to 23.4.1 and adds new snapshots
  to be compatible  with the new version.
- adds a script to run before every test that makes
  the test fail when there are console errors. This
  is helpful in catching missing propTypes from the
  tests, since those will print console errors, but
  not actually fail the test itself.
- Fixes PropType errors.
- Fixes unhandled promise warning.

johnpmitsch commented:
combined with https://github.com/Katello/katello/pull/7538, this should make our tests noise-free :sunglasses:

johnpmitsch commented on webpack/components/Search/index.js:
```diff
@@ -39,7 +39,7 @@ class Search extends Component {
         })),
       });
-    });
+    }).catch(() => {});
```
@tstrachota This is empty to prevent the unhandled promise warning, but do we want to catch errors here?

tstrachota replied:
If something breaks it would produce an error on each type in the field, right? In this case I'd avoid propagating the errors as toast notifications but I think it would make sense to log them into console.

tstrachota commented on webpack/global_test_setup.js:
```diff
@@ -0,0 +1,6 @@
+// output and traceback for actual error.
+global.console.error = (error) => {
+  throw new Error(error);
+};
```
I like it :+1:

johnpmitsch commented:
@tbrisker thanks, I guess I had something wrong with my environment, on a new environment I was able to reproduce the bad snapshot and corrected it (back to the original snapshot)

I modified the Search test to log the Promise errors to the console. To make sure nothing prints out during the test, I mocked the API.

I also added a `lint:test` script to package.json, since I get tired of running both lint and test :innocent:

johnpmitsch commented:
@tbrisker @waldenraines I've updated this and added fixes for the breakages in master. Since this is a small PR, I recommend we merge this if everyone is ok with it to get master branch back to a healthy state :thermometer: let me know your thoughts

waldenraines commented on webpack/components/Search/index.js:
```diff
@@ -33,13 +33,19 @@ class Search extends Component {
+      })
+      .catch((error) => {
+        // eslint-disable-next-line no-console
+        console.log(error);
```
Why do we want to `console.log()` this error?  We should do something useful with the error or do nothing at all.

johnpmitsch replied:
@waldenraines This is the question I asked @tbrisker [here](https://github.com/Katello/katello/pull/7555#discussion_r204158046), maybe he can expand on it? (the comment was collapsed so it was hidden).

johnpmitsch replied:
@waldenraines The reason I added the catch block is otherwise node throws a "unhandled promise" warning, but running this again, I see this is only because I was using a newer version of nodejs. I completely removed the catch block without any warnings, I think this is fine for now (until we update node versions)

tbrisker commented:
@johnpmitsch you seem to be mentioning me instead of @tstrachota ;)
